src/dsp_tools/utils/connection_live.py

```diff
@@ -1,12 +1,81 @@
 import json
+import time
 from dataclasses import dataclass
 from datetime import datetime
+from functools import partial
 from pathlib import Path
-from typing import Any, Optional, cast
+from typing import Any, Callable, Optional, cast
 
 import requests
+from requests import ReadTimeout, RequestException, Response
+from urllib3.exceptions import ReadTimeoutError
 
 from dsp_tools.models.exceptions import BaseError
+from dsp_tools.utils.create_logger import get_logger
+
+logger = get_logger(__name__)
+
+
+def _try_network_action(
+    action: Callable[..., Any],
+    *args: Any,
+    **kwargs: Any,
+) -> Any:
+    """
+    Helper method that tries 7 times to execute an action.
+    If a timeout error, a ConnectionError, a requests.exceptions.RequestException, or a non-permanent BaseError occors,
+    it waits and retries.
+    The waiting times are 1, 2, 4, 8, 16, 32, 64 seconds.
+    If another exception occurs, it escalates.
+
+    Args:
+        action: a lambda with the code to be executed, or a function
+        args: positional arguments for the action
+        kwargs: keyword arguments for the action
+
+    Raises:
+        BaseError: if the action fails permanently
+        unexpected exceptions: if the action fails with an unexpected exception
+
+    Returns:
+        the return value of action
+    """
+    action_as_str = f"{action=}, {args=}, {kwargs=}"
```

Comment:
not sure how much sense this makes. This is actually an unstructured log that corresponds 1:1 to the structured variant that is implemented in `_write_request_to_file()`. But I'm unsure how far we want to go in this PR, or if we want to improve this later.

Reply:
I agree that this doesn't make much sense as it is now, but I also think we should keep this for a future improvement. It doesn't make sense to log this each time we re-try either, as the information should stay the same with each attempt.